Comment with dependency diff

wzieba opened the pull request:
### Description

This PR introduces `comment_with_dependency_diff` script that is aiming to help Android product teams to better understand what changes to the dependency tree a PR might introduce.

We use similar scripts in WooCommerce and Day One projects. I want to extend support to other projects and share the implementation between them for easier maintenance.

Additionally, this PR introduces some improvements to the script:
- comment on PR without using Gradle (so our Android apps can remove another, not really needed dependency from their codebase)
- print build environment changes as well (previously, the script was only printing runtime dependencies changes)

To see the result, please see:
- Pocket Casts: https://github.com/Automattic/pocket-casts-android/pull/2857
- Matticspace Mobile (for the case of a large diff): https://github.com/Automattic/Matticspace-Mobile/pull/67

---

- [ ] I have considered if this change warrants release notes and have added them to the appropriate section in the `CHANGELOG.md` if necessary.

wzieba committed:
Raw copy of the script from Day One Android

wzieba committed:
Accept configuration, owner and repo parameters

wzieba committed:
Support Kotlin script (`*.gradle.kts`) files as well

wzieba committed:
Support Dependencies Catalog (`*.toml`) files

dangermattic commented:
<!--
  0 Errors
  1 Warning: Please add an entry in the <co...
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Please add an entry in the <code>CHANGELOG.md</code> file to describe the changes made by this PR</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

wzieba committed:
Use fastlane action for commenting on a PR

Instead of a Gradle plugin, which would have to be managed on project source code

wzieba committed:
Provide module argument

wzieba committed:
Remove dependencyTreeDiffCommentToGitHub execution if no .gradle files have changed

wzieba committed:
Fix checking for arguments

wzieba committed:
Use -E for file pattern matching

wzieba committed:
Fix typo when providing reuse-identifier

wzieba committed:
Fix running fastlane action

wzieba committed:
Add support for buildEnvironment diff

wzieba committed:
Use template to generate comment body

It's more readable this way

wzieba committed:
Actually create $DIFF_BUILD_ENV_FILE. Check if diff files exist before adding content to comment

wzieba committed:
Apply fixes to how comment body is created

wzieba committed:
Don't call `diff` tool

Interpret the content only as markdown

wzieba committed:
Improve formatting of the comment messages

wzieba committed:
Log status and diff before checkout

For debugging

wzieba committed:
Sometimes, a diff might be generated (e.g. Gemfile.lock after gem installation). We don't want to hold the script execution for this reason.

wzieba committed:
Use file to keep comment body

To maybe fix `line 129: /var/lib/buildkite-agent/.rbenv/shims/bundle: Argument list too long`

wzieba committed:
Another approach to fix Argument list too long

wzieba committed:
Verify if the problem is size of the comment

wzieba committed:
Use comment_on_pr script instead of fastlane action

wzieba committed:
Pass comment file to comment_on_pr

wzieba committed:
Possible fix for Argument list too long in comment_on_pr

wzieba committed:
Pass json_payload_file as a file

wzieba committed:
Pass raw file to jq

like in upload_sarif_to_github

wzieba committed:
Cover against GitHub comment max size

wzieba committed:
Check if adding the new content will exceed the maximum allowed size

Previously script would crash if project diff and build env diff had both e.g. 50k characters.

wzieba committed:
Improve markdown formatting

wzieba committed:
Remove quotation marks from configuration name

wzieba committed:
Don't require providing owner and repo names

Instead, get this from BUILDKITE env variable. I took this idea from comment_on_pr script.

wzieba committed:
Add a changelog entry

AliSoftware commented on bin/comment_on_pr:
```diff
@@ -126,20 +126,37 @@ $comment_body_id
+comment_body_file=$(mktemp)
+echo "$comment_body" > "$comment_body_file"
+
+# Use jq with the temporary file for payload
+#jq -c --null-input --arg body "$(cat "$comment_body_file")" '{body: $body}' > "$json_payload_file"
+
```
🧹  I guess this is leftover from debugging and can be removed?

AliSoftware commented on bin/comment_on_pr:
```diff
@@ -126,20 +126,37 @@ $comment_body_id
+# Use jq with the temporary file for payload
+#jq -c --null-input --arg body "$(cat "$comment_body_file")" '{body: $body}' > "$json_payload_file"
+
+json=$(jq -c --null-input \
+  --rawfile body "$comment_body_file" \
+  '{
+    body: $body
+   }
+  '
+)
+
+json_payload_file=$(mktemp)
+echo "$json" > "$json_payload_file"
```
💄  Alternative: avoid intermediate `json` variable altogether:

```suggestion
json_payload_file=$(mktemp)
jq -c --null-input \
  --rawfile body "$comment_body_file" \
  '{ body: $body }' \
  > "$json_payload_file"
```

AliSoftware commented on bin/comment_with_dependency_diff:
```diff
@@ -0,0 +1,153 @@
+#!/bin/bash
+
+set -euo pipefail
```
📚  The more we add utility scripts like those to this plugin, the more I realize we should have made a better job as providing help / documentation at what each of those do. Especially so that people not as familiar with our CI (e.g. devs that are interested in using some of those helper scripts in their pipelines themselves but have no idea what each utility does) can better understand the purpose of each.

For example, [the comment at the start of `comment_on_pr` is a good example of having some code-comment documentation on what the utility does and how it's expected to be used](https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/pull/120/files#diff-d536c5bec9be6d9bc743cc0175b25457a7417a97a3afded630db75505b91882cL3-R22)

I wish we took that habit sooner and applied it to all of our other `bin/*` utilities in this repo, making them all start with a code-comment explaining their role / context in which those helpers would be used and could help, documenting what the script does and its expected parameters and inputs (e.g. "this script should only run after `./gradlew …` has been called and generated the build artifacts", etc).
We should probably make a pass at all of them at some point.

In the meantime, wdyt about us already starting moving towards this good habit and add some intro doc comments in any new helpers we introduce, starting from this one?

AliSoftware commented on bin/comment_with_dependency_diff:
```diff
@@ -0,0 +1,153 @@
+DIFF_BUILD_ENV_FILE="$DIFF_DEPENDENCIES_FOLDER/diff_build_env.txt"
+DEPENDENCY_TREE_VERSION="1.2.1"
+
+PR_NUMBER=$BUILDKITE_PULL_REQUEST
```
💡 Even if one is only expected to call this helper on a Buildkite step for which we set `if: build.pull_request.id != null`, could be worth adding some protection about it anyway—typically at the start of the script, after your tests for `-z "$MODULE"` / `-z "$CONFIGURATION"`. Like:

```bash
if [ "${BUILDKITE_PULL_REQUEST:-false}" == "false" ]; then
  echo "This script should only be called on pull requests. Be sure to set `if: build.pull_request.id != null` on the step invoking it in your buildkite pipeline"
  exit 2
fi
```

AliSoftware commented on bin/comment_with_dependency_diff:
```diff
@@ -0,0 +1,153 @@
+
+echo "--> Starting the check"
+echo "--> Fetching the base branch from $PR_URL"
+githubResponse="$(curl "$PR_URL" -H "Authorization: token $GITHUB_TOKEN")"
```
💭  I have not checked if this would be strictly equivalent or not, but what do you think about using [our `github_api` helper from this repo](https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/blob/trunk/bin/github_api) instead of calling `curl` directly, now that this script lives in the same repo/plugin? 🤔

AliSoftware commented on bin/comment_with_dependency_diff:
```diff
@@ -0,0 +1,153 @@
+
+echo "--> Starting the check"
+echo "--> Fetching the base branch from $PR_URL"
+githubResponse="$(curl "$PR_URL" -H "Authorization: token $GITHUB_TOKEN")"
+baseBranch=$(echo "$githubResponse" | tr '\r\n' ' ' | jq '.base.ref' | tr -d '"')
+echo "--> Base branch is $baseBranch"
```
🤔  Actually, is the purpose of this whole API call only to extract the name of the base (aka target) branch of the PR?

If so, why not just use [`BUILDKITE_PULL_REQUEST_BASE_BRANCH`](https://buildkite.com/docs/pipelines/environment-variables#BUILDKITE_PULL_REQUEST_BASE_BRANCH) directly?

_(I think I slight remember mentioning this a while ago when I reviewed the first iterations of this script back when it was added to the repositories themselves, and iirc the argument was mostly due to legacy with this script previously being also used in other CI, maybe while we were still in the middle of transitioning to Buildkite…? So maybe that's where it comes from?)_

wzieba replied:
ah yes, thank you. It simplified things a lot!

AliSoftware commented on bin/comment_with_dependency_diff:
```diff
@@ -0,0 +1,153 @@
+echo "--> Downloading dependency-tree-diff.jar"
+curl -v -L "https://github.com/JakeWharton/dependency-tree-diff/releases/download/$DEPENDENCY_TREE_VERSION/dependency-tree-diff.jar" -o dependency-tree-diff.jar
+sha=$(sha1sum dependency-tree-diff.jar | cut -d' ' -f1)
+if [[ $sha != "39c50f22cd4211a3e52afec43120b7fdf90f9890" ]]; then
```
💡  Nitpick: could be worth extracting this SHA in a `DEPENDENCY_TREE_CHECKSUM` constant to put next to line 23, so that it'd be easier to update to a new version in the future by changing the version and checksum together?

AliSoftware commented on bin/comment_with_dependency_diff:
```diff
@@ -0,0 +1,153 @@
+    current_file_size=$(wc -c < "$COMMENT_FILE")
+
+    if (( current_file_size + ${#content} > github_pr_comment_max_size )); then
+      echo "Content exceeds $github_pr_comment_max_size characters. Navigate to Buildkite build artifacts to see details." >> $COMMENT_FILE
```
💡  I think we could even add a link to the job for quicker access (users would still need to click on the "Artifacts" tab but at least they'd already be on the right build and job 🙂 )
```suggestion
      echo "Content exceeds $github_pr_comment_max_size characters. [Navigate to Buildkite build artifacts](${BUILDKITE_BUILD_URL}#${BUILDKITE_JOB_ID}) to see details." >> $COMMENT_FILE
```

AliSoftware replied:
Btw:
 - There's a risk that `current_file_size` would be so close to `github_pr_comment_max_size` already that even just adding this "Content exceed …" line to it will make it go over the limit…
    - So the `github_pr_comment_max_size` limit value should really be 65536… minus the length of this line telling that the content exceeds the max size 😅 
 - Wouldn't the logic of `append_content` here add the `echo "Content exceeds …"` string multiple times if you call `append_content` multiple times and it's not just the last one but one before that that makes the length be exceeded?
    - Though I guess in this case it's not a big deal because you only call `append_content` a maximum number of 2 times, so this scenario can only happen if the content from `DIFF_DEPENDENCIES_FILE` is already too long, and in which case you'll print the `Content exceeds …` 2 times (one for when you call it for `DIFF_DEPENDENCIES_FILE`, then other when you call it for `DIFF_BUILD_ENV_FILE`)… so probably not that dramatic and not worth fixing I guess.
    - But just wanted to mention in case you wanted to look into it, or if you planned in future PRs to potentially add more info and call `append_content` more times in future iterations of this helper, just to keep that in mind.

AliSoftware reviewed (commented):
Looking good, just added some tweaking ideas.

wzieba committed:
Remove forgotten debug code

wzieba committed:
Improve clarity with json_payload_file creation

wzieba committed:
Add user manual for the comment_with_dependency_diff

wzieba committed:
Protect against lack of BUILDKITE_PULL_REQUEST

wzieba committed:
Use $BUILDKITE_PULL_REQUEST_BASE_BRANCH instead of fetching base branch from GH API

It simplifies things a lot!

wzieba committed:
Use uppercase for baseBranch

To match all other vars

wzieba committed:
Move dependency tree checksum

for easier updates in the future

wzieba committed:
Account the warnings message + header size when calculating possible comment size

wzieba committed:
Add a link to "content exceeds" message

wzieba commented:
Thank you @AliSoftware for the review! You raised good points. I think I've adjusted the implementation. Here are also updated tests:
- Matticspace Mobile (case with exceeding size: https://github.com/Automattic/Matticspace-Mobile/pull/67)
- Pocket Casts (casual case: https://github.com/Automattic/pocket-casts-android/pull/2857)

Both seem to still work fine.

wzieba commented:
Correction, title is missing

<img width="673" alt="image" src="https://github.com/user-attachments/assets/d75eb083-0ac8-49af-b542-3f3198fe8c45">

AliSoftware reviewed (approved):
:shipit:

wzieba committed:
Include header in "content exceeds" case

wzieba commented:
We're back on track!
<img width="953" alt="image" src="https://github.com/user-attachments/assets/040cbb21-3fa2-460a-81fa-812a002c5729">
https://github.com/Automattic/Matticspace-Mobile/pull/67

wzieba committed:
Merge branch 'trunk' into comment_with_dependency_diff

AliSoftware commented:
I just realized that there might be more to it to compute the exact comment body, because:
 - I'm guessing GitHub counts `\n` as a character (i.e. comparing with the current length + length of header + length of body is off by one or two due to the extra newlines)
 - `comment_on_pr` adds a some hidden content (`<!-- … -->` that includes the comment ID (`--id`), so that subsequent calls can from subsequent CI builds on the same PR can know which comment to edit / delete. So that adds a bit to the length as well

Maybe the easiest solution is to simply reduce the value of `github_comment_max_size` in your code to something like 65400, to give some margin for those extras characters and id comment?
